# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos:  Pedro Luis Soto Santos <br>
Url: https://github.com/plss12/03MAIR---Algoritmos-de-Optimizacion/blob/main/Trabajo/Seminario_Algoritmos.ipynb<br>

Problema:
>3. Combinar cifras y operaciones

Descripción del problema:

El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.

Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/).

Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad dada. Un ejemplo sería para obtener el 4:
4+2-6/3*1 = 4




                                        

In [14]:
from itertools import permutations

def numeros_posibles():
    numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    operadores = ['+', '-', '*', '/']
    numeros_expresiones = {}
    # Generar todas las permutaciones de 5 números y 4 operadores
    permuts_numeros = permutations(numeros, 5)
    for numeros_permutados in permuts_numeros:
        permuts_operadores = permutations(operadores, 4)
        for operadores_permutados in permuts_operadores:
            expresion = []
            
            # Alternar números y operadores
            for i in range(4):
                expresion.append(str(numeros_permutados[i]))
                expresion.append(operadores_permutados[i])
            expresion.append(str(numeros_permutados[4]))

            # Unir la expresión en una cadena
            expresion_str = ''.join(expresion)
            # Evaluar la expresión y guardar el resultado
            resultado = eval(expresion_str)
            # Si el resultado es un número entero y no se ha obtenido antes
            if resultado.is_integer() and resultado not in numeros_expresiones.keys():
                numeros_expresiones[resultado] = expresion_str
                
    return numeros_expresiones

# Cálculamos la lista completa de números enteros que se pueden obtener con las restricciones del problema
numeros = numeros_posibles()

print("Números enteros calculados:", len(numeros))

# Obetenemos el máximo y el mínimo de la lista, como se nos pide en el problema
num_max = max(numeros)
num_min = min(numeros)

print("Número entero Máximo:", num_max)
print("Número entero Mínimo:", num_min)

# A partir del máximo, el mínimo y la cantidad de números obtenidos, podemos comprobar si se pueden obtener
# todos los números enteros entre el mínimo y el máximo, como se nos pide en el problema
if num_max - num_min + 1 == len(numeros):
    print("Se pueden obtener todos los números enteros entre", num_min, "y", num_max)
else:
    print("No se pueden obtener todos los números enteros entre", num_min, "y", num_max)

Números enteros calculados: 147
Número entero Máximo: 77.0
Número entero Mínimo: -69.0
Se pueden obtener todos los números enteros entre -69.0 y 77.0


(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>



¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.




Si no tenemos en cuenta las restricciones tenemos que calcular el número total de posibles operaciones, sin tener en cuenta si el resultado de la operación es un número entero o un número racional. Para ello sería el resultado de calcular una variación sin repetición cogiendo 5 valores de los 9 posibles números, el resultado de este cálculo sería:

V(9,5) = 9!/(9-5)! = 15120

Para el caso de los signos nos encontramos con una permutación sin repetición:

P(4) = 4! = 24

Para obtener el número total de operaciones debemos multiplicar los dos anteriores resultados, 15120*24 = 362880. Este resultado ha sido calculado a priori y además ha sido calculado con el código de python de abajo.

En el caso de el número de posibilidades teniendo en cuenta las restricciones, debemos de contar solo las operaciones que dan como resultado un número entero y debemos de contar solo una operación por número entero, ya que no es necesario guardar dos operaciones que dan como resultado el mismo número entero, ya que el objetivo de este problema es obtener todos los números enteros que es posible encontrar con este tipo de operaciones. Para ello usamos la función que tenemos arriba en la que obtenemos todos los números enteros posibles de calcular con estas restricciones, siendo el resultado 147 operaciones o números.

In [12]:
def generar_combinaciones():
    numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    operadores = ['+', '-', '*', '/']
    expresiones = []
    # Generar todas las permutaciones de 5 números y 4 operadores
    permuts_numeros = permutations(numeros, 5)
    for numeros_permutados in permuts_numeros:
        permuts_operadores = permutations(operadores, 4)
        for operadores_permutados in permuts_operadores:
            expresion = []
            
            # Alternar números y operadores
            for i in range(4):
                expresion.append(str(numeros_permutados[i]))
                expresion.append(operadores_permutados[i])
            expresion.append(str(numeros_permutados[4]))

            # Unir la expresión en una cadena
            expresion_str = ''.join(expresion)
            expresiones.append(expresion_str)
    return expresiones
            
# Llamar a la función para generar y mostrar las combinaciones
combinaciones = generar_combinaciones()

print("Número de posibles operaciones sin restricciones:", len(combinaciones))

Número de posibles operaciones sin restricciones: 362880


Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


Se ha decidido utilizar una lista para almacenar los números y los operadores para poder ser usados por la función permutations. Para los resultados se ha optado por un diccionario, para almacenar cada número entero obtenido como clave, y acompañando a este número como valor se almacena la operación que da como resultado este número. Este diccionario hace muy sencillo el saber como se ha calculado cada número, ya que se podría almacenar únicamente una lista de números, pero al no tener su operación se perdería parte de la información que podría ser útil para comprobar el correcto funcionamiento del algoritmo.

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

Este problema no es un claro problema de optimización, o al menos yo no lo he planteado como tal, por lo que no creo que haya una función objetivo clara. Sin embargo se podría decir que la función objetivo es la maximicación de la cantidad de números enteros únicos obtenidos al evaluar las distintas expresiones obtenidas con las combinaciones. Por ello se podría decir que es un problema de maximicación, ya que buscamos maximizar la cantidad de números enteros calculados.

Diseña un algoritmo para resolver el problema por fuerza bruta

En este caso la función que resuelve el problema por fuerza bruta es el código utilizado al principio, lo vuelvo a poner a continuación.

In [40]:
def numeros_posibles():
    numeros = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    operadores = ['+', '-', '*', '/']
    numeros_expresiones = {}
    # Generar todas las permutaciones de 5 números y 4 operadores
    permuts_numeros = permutations(numeros, 5)
    for numeros_permutados in permuts_numeros:
        permuts_operadores = permutations(operadores, 4)
        for operadores_permutados in permuts_operadores:
            expresion = []
            
            # Alternar números y operadores
            for i in range(4):
                expresion.append(str(numeros_permutados[i]))
                expresion.append(operadores_permutados[i])
            expresion.append(str(numeros_permutados[4]))

            # Unir la expresión en una cadena
            expresion_str = ''.join(expresion)
            # Evaluar la expresión y guardar el resultado
            resultado = eval(expresion_str)
            # Si el resultado es un número entero y no se ha obtenido antes
            if resultado.is_integer() and resultado not in numeros_expresiones.keys():
                numeros_expresiones[resultado] = expresion_str
                
    return numeros_expresiones

# Cálculamos la lista completa de números enteros que se pueden obtener con las restricciones del problema
numeros = numeros_posibles()

print("Números enteros calculados:", len(numeros))


Números enteros calculados: 147


Calcula la complejidad del algoritmo por fuerza bruta

Este problema cuenta con 3 for anidados y un if para comprobar el resultado. El primer for corresponde las permutaciones de los números y el segundo a las permutaciones de los operadores, por lo que su complejidad es la formula de las permutaciones O(n!/(n-5)!) y O(n!/(n-4)!). El último for es sobre una constante, por lo que tiene una complejidad constante O(1). El if en el que se comprueba si el resultado es integer y si no está en el diccionario lo contaremos como complejidad constante O(1), aunque la comprobación de si el resultado está en el diccionario podría aumentar su complejidad dependiendo del total de registros que se guarden en este diccionario. Multiplicando estas complejidades obtenemos una complejidad de O(n!*m!/((n-5)!*(m-4)!)), siendo n la cantidad de números y m la cantidad de operadores utilizados.

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

(*)Calcula la complejidad del algoritmo

Respuesta

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Teniendo en cuenta que se utilizan permutaciones, no se puede utilizar un juego de datos que contengan datos repetidos, en el caso de los números y los operadores. También se debe tener en cuenta que para este problema la cantidad de números debe ser como mínimo mayor en una cantidad que el de operadores, o si no se podría obtener operaciones incorrectas. Por ello, creo que es posible aleatorizar el juego de datos utilizando una cantidad variable de números y operadores como podemos ver en el siguiente código.

In [20]:
cantidad_numeros = 7
cantidad_operadores = 3

Aplica el algoritmo al juego de datos generado

In [134]:

import random

def numeros_posibles(cantidad_numeros, cantidad_operadores):

    if cantidad_numeros > 9 or cantidad_numeros<cantidad_operadores:
        print("Debe haber como máximo 9 números y que la cantidad de números sea mayor que la de operadores")
    numeros = random.sample(range(1, 10), cantidad_numeros)
    operadores = random.sample(['+', '-', '*', '/'], k=cantidad_operadores)

    numeros_expresiones = {}
    # Generar todas las permutaciones de 5 números y 4 operadores
    cantidad_numeros_operacion = cantidad_operadores+1

    permuts_numeros = permutations(numeros, cantidad_numeros_operacion)
    for numeros_permutados in permuts_numeros:
        permuts_operadores = permutations(operadores, cantidad_operadores)
        for operadores_permutados in permuts_operadores:
            expresion = []
            
            # Alternar números y operadores
            for i in range(cantidad_operadores):
                expresion.append(str(numeros_permutados[i]))
                expresion.append(operadores_permutados[i])
            expresion.append(str(numeros_permutados[cantidad_operadores]))

            # Unir la expresión en una cadena
            expresion_str = ''.join(expresion)
            # Evaluar la expresión y guardar el resultado
            resultado = eval(expresion_str)
            
            # Si el resultado es un número entero y no se ha obtenido antes
            if float(resultado).is_integer() and resultado not in numeros_expresiones.keys():
                numeros_expresiones[resultado] = expresion_str
                
    return numeros_expresiones, numeros, operadores

# Cálculamos la lista completa de números enteros que se pueden obtener con las restricciones del problema
numeros, lista_numeros, lista_operadores = numeros_posibles(cantidad_numeros, cantidad_operadores)

print("Números utilizados:", lista_numeros)
print("Operadores utilizados:", lista_operadores)
print("Números enteros calculados:", len(numeros))

# Obetenemos el máximo y el mínimo de la lista, como se nos pide en el problema
num_max = max(numeros)
num_min = min(numeros)

print("Número entero Máximo:", num_max)
print("Número entero Mínimo:", num_min)

# A partir del máximo, el mínimo y la cantidad de números obtenidos, podemos comprobar si se pueden obtener
# todos los números enteros entre el mínimo y el máximo, como se nos pide en el problema
if num_max - num_min + 1 == len(numeros):
    print("Se pueden obtener todos los números enteros entre", num_min, "y", num_max)
else:
    print("No se pueden obtener todos los números enteros entre", num_min, "y", num_max)


Números utilizados: [1, 7, 4, 5, 3, 6, 2]
Operadores utilizados: ['*', '/', '-']
Números enteros calculados: 71
Número entero Máximo: 40.0
Número entero Mínimo: -40.0
No se pueden obtener todos los números enteros entre -40.0 y 40.0


Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Mayoritariamente he utilizado los apuntes obtenidos en la asignatura de Matemáticas para la Inteligencia Artificial para el temario de combinatoria, para poder realizar los cálculos a priori de la cantidad de combinaciones y corroborar luego que los datos obtenidos por el código de Python eran correctos.

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

La complejidad del problema es bastante alta, se me ocurren varias formas posibles de avanzar en el estudio del problema, las cúales llevarían bastante estudio. Como por ejemplo, el uso de poda para evitar evaluar ciertas ramas que a priori no serán buenos resultados teniendo en cuenta posibles operaciones que den resultados invalidos, como divisiones entre 0 o resultados decimales. También es cierto que el problema es algo ambiguo, no se aclara si se pide solo el máximo y el mínimo número entero calculado o se requiere la completitud de los números calculado, pero si es verdad que dependiendo de la variante se podría utilizar una heurística para, por ejemplo, ordenar los números y los operadores para obtener directamente los valores menores o mayores sin la necesidad de calcular todos ellos.